pkg/action/action.go

```diff
@@ -93,23 +101,45 @@ func (ac *ActionController) handle(ctx context.Context, action *storkv1.Action)
 	return nil
 }
 
-func printFunc(msg, stream string) {
-	switch stream {
-	case "out":
-		logrus.Infof(msg)
-	case "err":
-		logrus.Errorf(msg)
-	default:
-		logrus.Errorf("printFunc received invalid stream")
-		logrus.Errorf(msg)
+func (ac *ActionController) printFunc(action *storkv1.Action, reason string) func(string, string) {
+	return func(msg, stream string) {
+		actionLog := log.ActionLog(action)
+		switch stream {
+		case "out":
+			actionLog.Infof(msg)
+			ac.recorder.Event(action, v1.EventTypeNormal, reason, msg)
+		case "err":
+			actionLog.Errorf(msg)
+			ac.recorder.Event(action, v1.EventTypeWarning, reason, msg)
+		default:
+			actionLog.Errorf("printFunc received invalid stream")
+			actionLog.Errorf(msg)
+			ac.recorder.Event(action, v1.EventTypeWarning, reason, msg)
+		}
 	}
 }
 
 func (ac *ActionController) updateStatus(action *storkv1.Action, actionStatus storkv1.ActionStatus) {
 	action.Status = actionStatus
 	err := ac.client.Update(context.TODO(), action)
 	if err != nil {
-		logrus.Errorf("failed to update Action status %v/%v with error %v", action.Name, actionStatus, err)
+		log.ActionLog(action).Errorf("failed to update action status to %v with error %v", action.Status, err)
+	}
+}
+
+// delete any action older than actionExpiryTime
+func (ac *ActionController) pruneActions(namespace string) {
+	actionList, err := storkops.Instance().ListActions(namespace)
```

Comment:
Instead of calling ListActions you can prune the Action CR for which you originally got the reconcile event. 
Since you are requeueing every reconcile event, you should get an event for all Action CRs periodically. This will save us from making extra k8s API calls to ListActions on every reconcile period.

Reply:
That's a great idea!